contribute/design/1.x/ndm/20190705-ndm-disk-identification.md

```diff
@@ -0,0 +1,196 @@
+attaching to a different machine or a different SCSI port of the same machine.
+
+### Implementation Details/Notes/Constraints
+
+The implementation details for the disk identification. The current implementation 
+works well for physical disks which have a unique WWN and model assigned by the 
+manufacturer. But the implementation fails for `Virtual_disk`, because the fields
+are provider dependent and are not guaranteed to be unique. 
+
+#### Current Implementation
+The following data is fetched from the disk: 
+1. `ID_WWN`
+2. `ID_MODEL`
+3. `ID_SERIAL_SHORT`
+4. `ID_VENDOR`
+5. `ID_TYPE`
+
+If the `ID_MODEL` is not `Virtual_disk` or `EphemeralDisk`:
+	- field 1-4 are appended
+	- md5 hash of the result is generated which will be UID for the disk by NDM
+
+If the `ID_TYPE` is empty or `ID_MODEL` is `Virtual_disk` or `EphemeralDisk`:
+	- There is a chance that the values in fields 1-4 will be either empty or all 
+	  the same depending on the provider.
+	- fields 1-4 are appended. Along with it, the hostname and DEVNAME(`/dev/sda`) is
+	  appended to the result
+	- md5 hash of the result is generated which will be UID for the disk by NDM
+
+#### Shortcomings of current implementation
+In case of virtual disks, since we are using the DEVNAME for generating UID, everytime 
+the same disk comes at a different path, the UID gets changed. This results in getting 
+the device identified as a new disk. This will lead to data corruption as the disk that 
+was being claimed by some users has now came up with a new UID, but with the old data.
+Also, since this disk is now in an Unclaimed state, it will be available for others to
+claim possibly leading to data loss
+
+#### Proposed Implementation
+
+1. We try to generate the UUID using the current implementation. 
+
+2. If the `ID_TYPE` is empty or `ID_MODEL` is `Virtual_disk` or `EphemeralDisk` and `ID_WWN` is empty 
+
+    1. Check if the disk has a GPT label available on it.
+    	- If a GPT label is available we will use only that label for UID generation
+    	  of the disk. This will ensure that even if the path is changed, or the disk
+    	  comes up at a different node, the UID generated will be same and the disk
+    	  can be uniquely identified.
+    	- Once the GPT label is available, it is possible to identify partitions
+    	  also on the disk (if it exists). This will work by adding the partition
+    	  number along with the generated UID.
+    	  eg: disk-xxxx-1 (/dev/sdX1), disk-xxxx-2 (/dev/sdX2). 
+    	  This is guaranteed to be unique since the partition numbers won't change
+    	  even if the device path changes
+    2. Check if the disk has a Filesystem label available on it. This is in cases where the 
+       complete disk is formatted with a filesystem without a partition table. If filesystem
+       label is available we use it for generating the UUID.
+    3. If GPT and FS label is not available on the disk, we create a GPT partition table on the disk,
```

Comment:
This proposal looks good. Has the GPT partition creation been prototyped? What additional changes need to be made to the NDM DaemonSet. Also as a note, we may want to mention when the NDM DaemonSet is deleted, the partition created will remain as is.

Reply:
@kmova  The partition creation has not been prototyped yet.  For the NDM daemonset, whenever a UID generation fails, a create partition has to be invoked. We will not have to create UID for the partition in the same event, as it will come as a separate udev event.

Also, will mention that partition will remain on the disk, when daemonset is deleted as a drawback